server/controllers/prisonerScheduleController.ts

```diff
@@ -0,0 +1,126 @@
+import { RestClientBuilder } from '../data'
+import { PrisonApiClient } from '../data/interfaces/prisonApiClient'
+import { Prisoner } from '../interfaces/prisoner'
+import { mapHeaderNoBannerData } from '../mappers/headerMappers'
+import { formatName, groupBy, times } from '../utils/utils'
+import { formatDate } from '../utils/dateHelpers'
+import { NameFormatStyle } from '../data/enums/nameFormatStyle'
+import { GetEventScheduleItem } from '../interfaces/prisonApi/getEventScheduleItem'
+import { SelectedWeekDates } from '../interfaces/scheduledEvent'
+
+/**
+ * Parse requests for case notes routes and orchestrate response
+ */
+
+export default class PrisonerScheduleController {
+  private prisonApiClient: PrisonApiClient
+
+  constructor(private readonly prisonApiClientBuilder: RestClientBuilder<PrisonApiClient>) {}
+
+  public async displayPrisonerSchedule(req: Request, res: Response, prisonerData: Prisoner) {
+    const { firstName, middleNames, lastName } = prisonerData
+    const name = formatName(firstName, middleNames, lastName, { style: NameFormatStyle.firstLast })
+
+    const { clientToken } = res.locals
+    this.prisonApiClient = this.prisonApiClientBuilder(clientToken)
+
+    const selectedWeekDates: SelectedWeekDates[] = [] as SelectedWeekDates[]
+
+    let schedule: GetEventScheduleItem[] = [] as GetEventScheduleItem[]
+    const { when } = req.query
+
+    const { bookingId } = prisonerData
+
+    if (when === 'nextWeek') {
+      schedule = await this.prisonApiClient.getScheduledEventsForNextWeek(bookingId)
+    } else {
+      schedule = await this.prisonApiClient.getScheduledEventsForThisWeek(bookingId)
+    }
+
+    const groupedByDate = groupBy(schedule, 'eventDate')
+    const oneWeekToday = format(add(startOfToday(), { weeks: 1 }), 'yyyy-MM-dd')
+
+    if (when === 'nextWeek') {
+      times(7)((i: number) =>
+        selectedWeekDates.push({
+          date: format(add(startOfToday(), { weeks: 1, days: i }), 'yyyy-MM-dd'),
+        }),
+      )
+    } else {
+      times(7)((i: number) =>
+        selectedWeekDates.push({
+          date: format(add(startOfToday(), { days: i }), 'yyyy-MM-dd'),
+        }),
+      )
+    }
+
+    const filterMorning = (activities: GetEventScheduleItem[]) =>
+      activities && activities.filter(activity => getHours(parseInt(activity.startTime, 10)) < 12)
+
+    const filterAfternoon = (activities: GetEventScheduleItem[]) =>
+      activities && activities.filter(activity => getHours(parseInt(activity.startTime, 10)) < 17)
+
+    const filterEveningDuties = (activities: GetEventScheduleItem[]) =>
+      activities && activities.filter(activity => getHours(parseInt(activity.startTime, 10)) >= 17)
+
+    const byStartTimeThenByEndTime = (a: GetEventScheduleItem, b: GetEventScheduleItem) => {
+      isBefore(parseInt(a.startTime, 10), parseInt(b.startTime, 10))
+      if (isBefore(parseInt(a.startTime, 10), parseInt(b.startTime, 10))) return -1
```

Comment:
`a.startTime` looks like it's a full datetime from the mocks?
If so, `parseInt()` isn't going to work properly here.
`isBefore()` takes a `Date` object, so these could be something like `isBefore(new Date(a.startTime), new Date(b.startTime))`

Reply:
Thanks for taking a look. The isBefore takes a number or a date.

(alias) isBefore(date: number | Date, dateToCompare: number | Date): boolean
<img width="545" alt="Screenshot 2023-09-18 at 12 18 26" src="https://github.com/ministryofjustice/hmpps-prisoner-profile/assets/116157538/1dca6682-4d68-486e-94d1-e6b5c8aea3c3">

Reply:
Yeah, it takes either, but my concern was the `parseInt` bit - if `a.startTime` is just a numeric string, that's fine then, but if it's an ISO date (which it is in the mocks) like `2023-09-09T12:34:56` then `parseInt` will be "helpful" and just give you the year bit, which is not useful for sorting on the time.
```
parseInt('2023-09-09T12:34:56', 10) // 2023
```

Reply:
Ok, that makes sense - I have updated it.